spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java

```diff
@@ -0,0 +1,161 @@
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.boot.builder.SpringApplicationBuilder;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.context.EnvironmentAware;
+import org.springframework.core.env.Environment;
+import org.springframework.util.StringUtils;
+
+/**
+ * Default {@link BinderFactory} implementation.
+ *
+ * @author Marius Bogoevici
+ */
+public class DefaultBinderFactory<T> implements BinderFactory<T>, DisposableBean, EnvironmentAware {
+
+	private final Map<String, BinderConfiguration> binderConfigurations;
+
+	private final Map<String, BinderInstanceHolder<T>> binderInstanceCache = new HashMap<>();
+
+	private volatile Environment environment;
+
+	private volatile String defaultBinder;
+
+	public DefaultBinderFactory(Map<String, BinderConfiguration> binderConfigurations) {
+		this.binderConfigurations = new HashMap<>(binderConfigurations);
+	}
+
+	@Override
+	public void setEnvironment(Environment environment) {
+		this.environment = environment;
+	}
+
+	public void setDefaultBinder(String defaultBinder) {
+		this.defaultBinder = defaultBinder;
+	}
+
+	@Override
+	public void destroy() throws Exception {
+		for (Map.Entry<String, BinderInstanceHolder<T>> entry : binderInstanceCache.entrySet()) {
+			BinderInstanceHolder<T> binderInstanceHolder = entry.getValue();
+			binderInstanceHolder.getBinderContext().close();
+		}
+	}
+
+	@Override
+	public synchronized Binder<T> getBinder(String name) {
+		String configurationName;
+		// Fall back to a default if no argument is provided
+		if (StringUtils.isEmpty(name)) {
+			if (binderConfigurations.size() == 0) {
+				throw new IllegalStateException("A default binder has been requested, but there there is no binder available");
+			}
+			else if (binderConfigurations.size() == 1) {
+				configurationName = binderConfigurations.keySet().iterator().next();
+			}
+			else {
+				if (StringUtils.hasText(defaultBinder)) {
+					configurationName = defaultBinder;
+				}
+				else {
+					throw new IllegalStateException(
+							"A default binder has been requested, but there is more than one binder available: "
+									+ StringUtils.collectionToCommaDelimitedString(binderConfigurations.keySet()) + ", and"
+									+ " no default binder has been set.");
+				}
+			}
+		} else {
+			configurationName = name;
+		}
+		if (!binderInstanceCache.containsKey(configurationName)) {
+			BinderConfiguration binderConfiguration = binderConfigurations.get(configurationName);
+			if (binderConfiguration == null) {
+				throw new IllegalStateException("Unknown binder configuration: " + configurationName);
+			}
+			Properties binderProperties = binderConfiguration.getProperties();
+			// Convert all properties to arguments, so that they receive maximum precedence
+			ArrayList<String> args = new ArrayList<>();
+			for (Map.Entry<Object, Object> property : binderProperties.entrySet()) {
```

Comment:
...or should `BinderConfiguration` itself initialize the `Properties` in its constructor (empty if none provided)? That way it would never be `null`.